quantfinlib/train_test_split/_base.py

```diff
@@ -0,0 +1,175 @@
+        if not groups.is_monotonic_increasing:
+            raise ValueError("groups must be sorted in ascending order")
+    else:
+        raise TypeError(
+            "delta_t must be an integer and groups must be a numpy array, \
+            or delta_t must be a Timedelta and groups must be a DatetimeIndex"
+        )
+
+
+def _purge(
+    train_index: np.ndarray,
+    test_index: np.ndarray,
+    indices: np.ndarray,
+    groups: Union[np.ndarray, pd.Series, pd.DatetimeIndex],
+    n_purge: Union[Integral, pd.Timedelta],
+) -> np.ndarray:
+    """Remove training indices whose groups fall within `n_purge` from the last test group.
+
+    Parameters
+    ----------
+    train_index : np.ndarray
+        The initial training indices.
+    test_index : np.ndarray
+        The test indices.
+    indices : np.ndarray
+        All indices.
+    groups : Union[np.ndarray, pd.Series, pd.DatetimeIndex]
+        The groups
+    n_purge : Union[int, pd.Timedelta]
+        Purge period.
+
+    Returns
+    -------
+    np.ndarra
+        The purged training indices.
+
+    Raises
+    ------
+    ValueError
+        If all training indices were purged
+    """
+    _validate_purge_embargo_inputs(train_index, test_index, indices, groups, n_purge)
+    max_test_group = groups[test_index].max()
+    purged_index = indices[(groups <= max_test_group + n_purge) & (groups > max_test_group)]
+    train_index = np.setdiff1d(train_index, purged_index)
+    if len(train_index) == 0:
+        raise ValueError("All training indices were purged. No training data available.")
+    return np.setdiff1d(train_index, purged_index)
+
+
+def _embargo(
+    train_index: np.ndarray,
+    test_index: np.ndarray,
+    indices: np.ndarray,
+    groups: Union[np.ndarray, pd.Series, pd.DatetimeIndex],
+    n_embargo: Union[Integral, pd.Timedelta],
+) -> np.ndarray:
+    """Remove training indices whose groups fall within `n_embargo` from the first test group.
+
+    Parameters
+    ----------
+    train_index : np.ndarray
+        The initial training indices.
+    test_index : np.ndarray
+        The test indices.
+    indices : np.ndarray
+        All indices.
+    groups : Union[np.ndarray, pd.Series, pd.DatetimeIndex]
+        The groups
+    n_embargo : Union[int, pd.Timedelta]
+        Embargo period.
+
+    Returns
+    -------
+    np.ndarray
+        The embargoed training indices.
+
+    Raises
+    ------
+    ValueError
+        If all training indices were embargoed.
+    """
+    _validate_purge_embargo_inputs(train_index, test_index, indices, groups, n_embargo)
+    min_test_group = groups[test_index].min()
+    embargoed_index = indices[(groups >= min_test_group - n_embargo) & (groups < min_test_group)]
+    train_index = np.setdiff1d(train_index, embargoed_index)
+    if len(train_index) == 0:
+        raise ValueError("All training indices were embargoed. No training data available.")
+    return train_index
+
+
+class BaseCV(ABC):

```

Comment:
would be super cool to somehow visualize the split result. to see what was embargoed, purged, etc. because you yield a generator and not full result at once, it may not be feasible until fully yielded, I am not sure.

Reply:
Added an example notebook.